readme.md

```diff
@@ -8,30 +8,100 @@
-While you’re still on remark 12, use the `gfm` option for remark.
-Use this plugin for remark 13+.
+**[remark][]** plugin to support [GFM][] (autolink literals, footnotes,
+strikethrough, tables, tasklists).
+
+## Contents
+
+*   [What is this?](#what-is-this)
+*   [When should I use this?](#when-should-i-use-this)
+*   [Install](#install)
+*   [Use](#use)
+*   [API](#api)
+    *   [`unified().use(remarkGfm[, options])`](#unifieduseremarkgfm-options)
+*   [Examples](#examples)
+    *   [Example: `singleTilde`](#example-singletilde)
+    *   [Example: `stringLength`](#example-stringlength)
+*   [Types](#types)
+*   [Compatibility](#compatibility)
+*   [Security](#security)
+*   [Related](#related)
+*   [Contribute](#contribute)
+*   [License](#license)
+
+## What is this?
+
+This package is a [unified][] ([remark][]) plugin to enable the extensions to
+markdown that GitHub adds: autolink literals (`www.x.com`), footnotes (`[^1]`),
+strikethrough (`~~stuff~~`), tables (`| cell |…`), and tasklists (`* [x]`).
+You can use this plugin to add support for parsing and serializing them.
+These extensions by GitHub to CommonMark are called [GFM][] (GitHub Flavored
+Markdown).
+
+**unified** is a project that transforms content with abstract syntax trees
+(ASTs).
+**remark** adds support for markdown to unified.
+**mdast** is the markdown AST that remark uses.
+This is a remark plugin that transforms mdast.
+
+## When should I use this?
+
+This project is useful when you want to support the same features that GitHub
+does in files in a repo, Gists, and several other places.
+Users frequently believe that some of these extensions, specifically autolink
+literals and tables, are part of normal markdown, so using `remark-gfm` will
+help match your implementation to their understanding of markdown.
+There are several edge cases where GitHub’s implementation works in unexpected
+ways or even different than described in their spec, so *writing* in GFM is not
+always the best choice.
+
+This plugin does not handle how markdown is turned to HTML.
+That’s done by [`remark-rehype`][remark-rehype].
+If your content is not in English and uses footnotes, you should configure that
+plugin.
+When generating HTML, you might also want to enable [`rehype-slug`][rehype-slug]
+to add `id`s on headings.
+
+A different plugin, [`remark-frontmatter`][remark-frontmatter], adds support for
+frontmatter.
+GitHub supports YAML frontmatter for files in repos and Gists but they don’t
+treat it as part of GFM.
+
+Another plugin, [`remark-github`][remark-github], adds support for how markdown
+works in relation to a certain GitHub repo in comments, issues, PRs, and
+releases, by linking references to commits, issues, and users.
+
+Yet another plugin, [`remark-breaks`][remark-breaks], turns soft line endings
+(enters) into hard breaks (`<br>`s).
+GitHub does this in a few places (comments, issues, PRs, and releases), but it’s
+not semantic according to HTML.
```

Comment:
> but it’s not semantic according to HTML

I'm not sure I follow this part, markdown isn't HTML, it can follow different rules.

Reply:
it relates to now how the plugin works (on the markdown AST), but what it entails when going to HTML (as described in the previous sentence).
Turning any line break into a `<br>` is not semantic per HTML

Reply:
I'm still confused.
Yes in HTML itself `\n` and `<br>` mean two different things.
But the way you phrase this, it sounds like you are referring to how the content is authored, markdown, where HTML semantics don't really apply.

In addition adding `<br>` where you want a break in HTML, is semantic HTML, and that is what this plugin does?

Reply:
I don’t get your first point, about relating to authoring.

> Yet another plugin […] turns soft line endings (enters) into hard breaks (`<br>`s). GitHub does this in a few places […] but it’s not semantic according to HTML.

To me, the `this` and `it` can only relate to what the plugin does, as I see no reference in that sentence to authoring.

> In addition adding `<br>` where you want a break in HTML, is semantic HTML, and that is what this plugin does?

In most cases, no. Only in poems and addresses. https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-br-element, https://developer.mozilla.org/en-US/docs/Web/HTML/Element/br. While possible, I think it’s quite unlikely that users only use soft line breaks in markdown paragraphs when in a poem or address, and leave any other paragraph without any enters in them.

Reply:
> > Yet another plugin […] turns soft line endings (enters) into hard breaks (<br>s). GitHub does this in a few places […] but it’s not semantic according to HTML.
>
> To me, the this and it can only relate to what the plugin does, as I see no reference in that sentence to authoring.

What I read:
I write Markdown on GitHub -> GitHub adds breaks -> The breaks in HTML are not Semantic HTML
That makes no sense, I intended the breaks, the plugin added the breaks, the `<br>` tag is intended to be used for content breaks.

> In most cases, no. Only in poems and addresses

:thinking: I dunno about that

> br elements must be used only for line breaks that are _actually part of the content_

In GFM semantics, breaks are part of the content, ergo, when converted to HTML the use of `<br>` is semantic.

Reply:
Perhaps HTML can be left out of this section entirely.

```md
Yet another plugin, [`remark-breaks`][remark-breaks], turns soft line endings (`\n`) into hard breaks (`\\\n` or `··\n`).
GitHub does this in a few places (comments, issues, PRs, and releases), but it’s not semantic according to CommonMark.
```

As far as I can tell, [what’s what `remark-breaks` really does](https://github.com/remarkjs/remark-breaks/blob/main/index.js#L34) anyway.

Reply:
> What I read: I write Markdown on GitHub -> GitHub adds breaks -> The breaks in HTML are not Semantic HTML

This is the intended reading of that sentence.

> I intended the breaks

Intending a `center` tag to be there doesn’t make it semantic. Semantics don’t come from imagining something into existence, semantics come from HTML.

> I dunno about that

I linked to the spec, which says this, and MDN, which explains it in more words. More examples can be found by searching the web, e.g., https://stackoverflow.com/questions/1726073/is-it-sometimes-bad-to-use-br, even https://www.w3schools.com/tags/tag_br.asp.

> In GFM semantics, breaks are part of the content, ergo, when converted to HTML the use of `<br>` is semantic.

This is incorrect on several points:

- GFM does not describe semantics, neither does CommonMark, HTML does
- GFM does not describe soft line breaks or `br`s, CommonMark does
- GFM does not describe this behavior that GitHub applies to comments, issue/PRs, and releases, and `remark-breaks`

---

What `remark-breaks` does is match one of the behaviors of github.com.
`<br>`s are highly unlikely to ever be semantic. They’re semantic in very few places. GH removes that deliberation. There is no choice. Do I understand it? Yes. Is it semantic? No.

Reply:
@remcohaszing the behavior of `remark-breaks` is fine per commonmark, as CM doesn’t define anything about semantics. Semantics come into play when considering HTML. And that’s to my knowledge the only reason why `remark-breaks` is ever used: to turn `\n` into an HTML `<br>` in the same way that GitHub does.

Reply:
> Semantics don’t come from imagining something into existence, semantics come from HTML.

Markdown is not HTML

> GFM does not describe semantics, neither does CommonMark, HTML does

I'd argue that it does.
And even if it doesn't, that does _not_ mean HTML semantics can be thrown at it, it isn't HTML.

> `<br>`s are highly unlikely to ever be semantic. They’re semantic in very few places. GH removes that deliberation. There is no choice. Do I understand it? Yes. Is it semantic? No.

This sounds like a complaint to take to the GitHub platform issue tracker or add to an FAQ in `remark-breaks` itself, it feels very off to have it here.

> GFM does not describe this behavior that GitHub applies to comments, issue/PRs, and releases, and remark-breaks

GFM does describe this behavior, in code which is not public.
It is not in the spec yet, but again, that is something to take up with GitHub.

---

I'd second @remcohaszing's recommendation https://github.com/remarkjs/remark-gfm/pull/25#discussion_r741128366 removing the HTML comment entirely.

I still don't think it is necessarily accurate.
Even if it is, it feels off to take a potshot at a feature in GitHub you disagree with in a readme like this, it adds more confusion than clarifying anything for a newcomer.

```suggestion
GitHub does this in a few places (comments, issues, PRs, and releases).
```

Reply:
> And even if it doesn't, that does not mean HTML semantics can be thrown at it, it isn't HTML.

People use that plugin to generate HTML.

> or add to an FAQ in remark-breaks itself

Already there.

> Even if it is, it feels off to take a potshot at a feature in GitHub you disagree with in a readme like this, it adds more confusion than clarifying anything for a newcomer.

This section is titled “When should I use this?”. It explains when (not) to use this and other plugins.
Users use `remark-breaks` because they want to turn every soft line break in markdown into a `<br>` in HTML. That is wrong.

---

Aside: I find it extremely important to educate newcomers when they’re f-in up the web.
It is extremely important that markdown processors used by millions on billions of pages are held to the highest of high standards.

Reply:
> Aside: I find it extremely important to educate newcomers when they’re f-in up the web.
It is extremely important that markdown processors used by millions on billions of pages are held to the highest of high standards.

Why not add a "semantic" alongside a "github" mode to `remark-breaks`, where `semantic` mode produces HTML that doesn't mess up the web?

Reply:
How would that mode work?

I don’t think a plugin can. Take:

```markdown
Here is some long
text.

You can send the package to:

Lorem Ipsumstreet 123
1234 AB City Name
Some country
```

`remark-breaks` adds a `<br>` between `long` and `text`, and two in the address.
The proper way to mark that up through markdown for HTML is with normal markdown without `remark-breaks`:

```markdown
Here is some long
text.

You can send the package to:

Lorem Ipsumstreet 123\
1234 AB City Name\
Some country
```